File: frontend/src/api/review.ts
import { API_URL, ApiError, request } from "./http";

// ─── Types ──────────────────────────────────────────────────────────────────

export interface ConflictListItem {
  conflict_id: string;
  comment_id: string;
  dataset_id: string;
  dataset_name: string;
  author_display_name: string;
  text_original: string;
  label_a: string;
  annotator_a: string;
  justificativa_a: string | null;
  label_b: string;
  annotator_b: string;
  justificativa_b: string | null;
  status: string;
  created_at: string;
}

export interface ConflictComment {
  comment_db_id: string;
  text_original: string;
  like_count: number;
  reply_count: number;
  published_at: string;
}

export interface AnnotationSide {
  annotator: string;
  label: string;
  justificativa: string | null;
  annotated_at: string;
}

export interface ConflictDetail {
  conflict_id: string;
  status: string;
  dataset_name: string;
  author_channel_id: string;
  author_display_name: string;
  comments: ConflictComment[];
  annotation_a: AnnotationSide;
  annotation_b: AnnotationSide;
  resolved_by: string | null;
  resolved_label: string | null;
  resolved_at: string | null;
}

export interface ResolveResponse {
  conflict_id: string;
  status: string;
  resolved_label: string;
  resolved_by: string;
  resolved_at: string;
}

export interface BotAnnotationDetail {
  annotator_name: string;
  label: string;
  justificativa: string | null;
}

export interface BotCommentItem {
  comment_db_id: string;
  text_original: string;
  author_display_name: string;
  author_channel_id: string;
  dataset_id: string;
  dataset_name: string;
  annotations: BotAnnotationDetail[];
  has_conflict: boolean;
  conflict_id: string | null;
}

export interface PaginatedResponse<T> {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  items: T[];
}

export interface ReviewStats {
  total_conflicts: number;
  pending_conflicts: number;
  resolved_conflicts: number;
  total_bots_flagged: number;
}

export interface ImportResult {
  imported: number;
  skipped: number;
  errors: string[];
}

// ─── API ────────────────────────────────────────────────────────────────────

export const reviewApi = {
  listConflicts: (
    token: string,
    params?: {
      status?: string;
      video_id?: string;
      dataset_id?: string;
      page?: number;
      page_size?: number;
    }
  ) => {
    const qs = new URLSearchParams();
    if (params?.status) qs.set("status", params.status);
    if (params?.video_id) qs.set("video_id", params.video_id);
    if (params?.dataset_id) qs.set("dataset_id", params.dataset_id);
    qs.set("page", String(params?.page ?? 1));
    qs.set("page_size", String(params?.page_size ?? 20));
    return request<PaginatedResponse<ConflictListItem>>(`/review/conflicts?${qs}`, {}, token);
  },

  getConflict: (conflictId: string, token: string) =>
    request<ConflictDetail>(`/review/conflicts/${conflictId}`, {}, token),

  resolve: (data: { conflict_id: string; resolved_label: "bot" | "humano" }, token: string) =>
    request<ResolveResponse>(
      "/review/resolve",
      {
        method: "POST",
        body: JSON.stringify(data),
      },
      token
    ),

  listBots: (
    token: string,
    params?: { video_id?: string; dataset_id?: string; page?: number; page_size?: number }
  ) => {
    const qs = new URLSearchParams();
    if (params?.video_id) qs.set("video_id", params.video_id);
    if (params?.dataset_id) qs.set("dataset_id", params.dataset_id);
    qs.set("page", String(params?.page ?? 1));
    qs.set("page_size", String(params?.page_size ?? 20));
    return request<PaginatedResponse<BotCommentItem>>(`/review/bots?${qs}`, {}, token);
  },

  stats: (token: string) => request<ReviewStats>("/review/stats", {}, token),

  importReview: async (
    data: {
      dataset_name: string;
      video_id: string;
      comments: Array<{
        comment_db_id: string;
        author_channel_id: string;
        author_display_name: string;
        text_original: string;
        final_label: "bot" | "humano";
        annotations?: Array<{
          annotator: string;
          label: string;
          justificativa?: string | null;
        }>;
        resolution?: {
          resolved_by?: string;
          resolved_label: string;
          resolved_at?: string;
        } | null;
      }>;
    },
    token: string,
    onProgress?: (sent: number, total: number) => void
  ): Promise<ImportResult> => {
    const CHUNK_SIZE = 2000;
    const all = data.comments;
    const total = all.length;
    const firstBatch = all.slice(0, CHUNK_SIZE);
    const hasMore = total > CHUNK_SIZE;

    const result = await request<ImportResult>(
      "/review/import",
      {
        method: "POST",
        body: JSON.stringify({
          ...data,
          comments: firstBatch,
          done: !hasMore,
        }),
      },
      token
    );
    onProgress?.(firstBatch.length, total);

    if (!hasMore) return result;

    let offset = CHUNK_SIZE;
    let totalImported = result.imported;

    while (offset < total) {
      const chunk = all.slice(offset, offset + CHUNK_SIZE);
      const isLast = offset + chunk.length >= total;
      const chunkResult = await request<{
        total_imported: number;
        chunk_received: number;
        done: boolean;
      }>(
        "/review/import-chunk",
        {
          method: "POST",
          body: JSON.stringify({ comments: chunk, done: isLast }),
        },
        token
      );
      totalImported += chunkResult.total_imported;
      offset += chunk.length;
      onProgress?.(offset, total);
    }

    return {
      imported: totalImported,
      skipped: result.skipped,
      errors: result.errors,
    };
  },

  downloadExport: async (
    format: "json" | "csv",
    token: string,
    datasetId: string
  ): Promise<void> => {
    const qs = new URLSearchParams();
    qs.set("format", format);
    qs.set("dataset_id", datasetId);

    const res = await fetch(`${API_URL}/review/export?${qs}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) {
      const body = (await res.json().catch(() => ({}))) as { detail?: string };
      throw new ApiError(body.detail ?? "Erro ao exportar revisão.", res.status);
    }

    // Use filename from Content-Disposition if available
    const cd = res.headers.get("Content-Disposition") ?? "";
    const filenameMatch = cd.match(/filename="(.+?)"/);
    const filename = filenameMatch ? filenameMatch[1] : `review.${format}`;

    const blob = await res.blob();
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    a.click();
    URL.revokeObjectURL(url);
  },
};

File: backend/schemas/review.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# ─── Request ─────────────────────────────────────────────────────────────────


class ResolveRequest(BaseModel):
    conflict_id: uuid.UUID
    resolved_label: Literal["bot", "humano"]


class ReviewImportComment(BaseModel):
    comment_db_id: uuid.UUID
    author_channel_id: str
    author_display_name: str
    text_original: str
    final_label: Literal["bot", "humano"]
    annotations: list[dict] = []
    resolution: dict | None = None


class ReviewImport(BaseModel):
    """Aceita o mesmo formato JSON exportado pelo export."""

    dataset_name: str
    video_id: str
    comments: list[ReviewImportComment] = Field(min_length=1)
    done: bool = True


class ReviewImportChunk(BaseModel):
    """Batch adicional de comentários revisados para import paginado."""

    comments: list[ReviewImportComment] = Field(min_length=1)
    done: bool = False


# ─── Response ────────────────────────────────────────────────────────────────


class ConflictListItem(BaseModel):
    conflict_id: uuid.UUID
    comment_id: uuid.UUID
    dataset_id: uuid.UUID
    dataset_name: str
    author_display_name: str
    text_original: str
    label_a: str
    annotator_a: str
    justificativa_a: str | None
    label_b: str
    annotator_b: str
    justificativa_b: str | None
    status: str
    created_at: datetime


class AnnotationSide(BaseModel):
    annotator: str
    label: str
    justificativa: str | None
    annotated_at: datetime


class ConflictComment(BaseModel):
    comment_db_id: uuid.UUID
    text_original: str
    like_count: int
    reply_count: int
    published_at: datetime


class ConflictDetail(BaseModel):
    conflict_id: uuid.UUID
    status: str
    dataset_name: str
    author_channel_id: str
    author_display_name: str
    comments: list[ConflictComment]
    annotation_a: AnnotationSide
    annotation_b: AnnotationSide
    resolved_by: str | None
    resolved_label: str | None
    resolved_at: datetime | None


class ResolveResponse(BaseModel):
    conflict_id: uuid.UUID
    status: str
    resolved_label: str
    resolved_by: str
    resolved_at: datetime


class BotAnnotationDetail(BaseModel):
    annotator_name: str
    label: str
    justificativa: str | None


class BotCommentItem(BaseModel):
    comment_db_id: uuid.UUID
    text_original: str
    author_display_name: str
    author_channel_id: str
    dataset_id: uuid.UUID
    dataset_name: str
    annotations: list[BotAnnotationDetail]
    has_conflict: bool
    conflict_id: uuid.UUID | None


class ReviewStats(BaseModel):
    total_conflicts: int
    pending_conflicts: int
    resolved_conflicts: int
    total_bots_flagged: int


class ImportResult(BaseModel):
    imported: int
    skipped: int
    errors: list[str]


class ImportChunkResponse(BaseModel):
    total_imported: int
    chunk_received: int
    done: bool
